src/data/nat/basic.lean

```diff
@@ -1373,6 +1386,28 @@ by rw [nat.mul_add_mod, nat.mod_eq_of_lt h]
 lemma pred_eq_self_iff {n : ℕ} : n.pred = n ↔ n = 0 :=
 by { cases n; simp [(nat.succ_ne_self _).symm] }
 
+/-- Recursion principle based on the div and modulus decomposition against some base `b : ℕ`:
+given a map `C (b * k + r) → C (b * (k + 1) + r)` for each `r < b` and `k`,
+and the construction of `C r` for all `r < b`, one can construct `C n` for any `n : ℕ`.  -/
+def div_mod_induction {C : ℕ → Sort*} (n b : ℕ) (hb : b ≠ 0) (hr : ∀ r < b, C r)
+  (hm : ∀ (k : ℕ) (r < b), C (b * k + r) → C (b * (k + 1) + r)) : C n :=
```

Comment:
Does this lose any generality as
```suggestion
  (hm : ∀ (k : ℕ), C (k) → C (k + b)) : C n :=
```
That's the version I feel like we already have

Reply:
Why is having this looser version better? It seems to me that it's a harder hypothesis to prove for downstream users of this recursion principle -- they have to prove it over all `C k`. With the one I had, the responsibility is only on the `r [MOD b]` congruence classes.

Reply:
Did this change make the other proof in this PR easier or harder?

Reply:
It simplified the other proof, thank you.